### PR TITLE
[Asset] Grid Improvements - Batch edit Asset metadata

### LINKED ISSUE #1662: Asset folder -> List - add filter for Metadata
## Feature description 
If you click on a asset folder you can view the assets of the folder in the "List" tab. 
It would be great if a user could filter for asset metadata in this view.

A use case for this could be: 
An asset is valid from day x to y. The user want's to filter all assets which are invalid.

Things to consider:
- Column selection (which metadatafields should be displayed)
- Saving column configuration (similar to object grids)
- Localized metadata (language selection)
- Editing Metadata in the grid view
- Show thumnails of the asset in the grid view
- CSV Export ?

### PR DESCRIPTION
**Pull request by dvesh3**
## Changes in this pull request  
Resolves #1662

## Additional info  
This is dependent of #5093

### CODE REVIEW

**Comment by kubaplas** [comment]
@dvesh3 Nice work! As the code is similar to one in DataObject controllers, do you have plans to extract the common base to separate services or helpers?

@brusch as I've been looking into moving out some code from Controller is there a preffered place for code (sql, logic) out of them?

**Comment by dvesh3** [comment]
Hi @kubaplas thanks for your review! we are trying our best to refactor the common functionality as it also leads to testing DataObject grid,..however it will be great if you can review it again after we merge this one and provide new PR for your changes :)

For SQL queries, I would say preferred way is to go with Dao classes.

**Comment by weisswurstkanone** [comment]
what happens if one creates a predefined metadata definition out of the set of default metadata fields (copyright, alt, title) ?

image

image

File: bundles/AdminBundle/Resources/public/js/pimcore/asset/helpers/gridCellEditor.js [hunk 1]
```diff
@@ -0,0 +1,163 @@
+/**
```
Comment (weisswurstkanone) [inline bundles/AdminBundle/Resources/public/js/pimcore/asset/helpers/gridCellEditor.js]: the whole file seems to be 95% identical to the object's grid cell editor

File: bundles/AdminBundle/Resources/public/js/pimcore/asset/listfolder.js [hunk 2]
```diff
@@ -14,7 +14,7 @@
 pimcore.registerNS("pimcore.asset.listfolder");
 pimcore.asset.listfolder = Class.create(pimcore.asset.helpers.gridTabAbstract, {
 
-    systemColumns: ["id", "type", "fullpath", "filename", "creationDate", "modificationDate", "preview", "size"],
+    systemColumns: ["id~system", "type~system", "fullpath~system", "filename~system", "creationDate~system", "modificationDate~system", "preview~system", "size~system"],
```
Comment (weisswurstkanone) [inline bundles/AdminBundle/Resources/public/js/pimcore/asset/listfolder.js]: please create a seperate issue for disallowing ~ in the metadata key

File: models/Asset/Service.php [hunk 3]
```diff
@@ -185,6 +185,7 @@ public static function gridAssetData($asset, $fields = null, $requestedLanguage
 
         if ($asset instanceof Asset && !empty($fields)) {
             $data = [
+                'id' => $asset->getid(),
```
Comment (weisswurstkanone) [inline models/Asset/Service.php]: ->getId()

**Comment by dvesh3** [comment]
Predefined Metadata will not show up in grid config if it has same key as Default Metadata.
> what happens if one creates a predefined metadata definition out of the set of default metadata fields (copyright, alt, title) ?
> 
> image
> 
> image

**Pull request merged**